dotnet-sourcelink-shared/SourceLinkMap.cs

```diff
@@ -0,0 +1,198 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+// TODO: Use a source package once available (https://github.com/dotnet/sourcelink/issues/443)
+
+namespace Microsoft.SourceLink.Tools
+{
+    internal sealed class SourceLinkMap
+    {
+        private readonly List<(FilePathPattern key, UriPattern value)> _entries;
+
+        public SourceLinkMap(List<(FilePathPattern key, UriPattern value)> entries)
+        {
+            Debug.Assert(entries != null);
+            _entries = entries;
+        }
+
+        internal struct FilePathPattern
+        {
+            public readonly string Path;
+            public readonly bool IsPrefix;
+
+            public FilePathPattern(string path, bool isPrefix)
+            {
+                Debug.Assert(path != null);
+
+                Path = path;
+                IsPrefix = isPrefix;
+            }
+        }
+
+        internal struct UriPattern
+        {
+            public readonly string Prefix;
+            public readonly string Suffix;
+
+            public UriPattern(string prefix, string suffix)
+            {
+                Debug.Assert(prefix != null);
+                Debug.Assert(suffix != null);
+
+                Prefix = prefix;
+                Suffix = suffix;
+            }
+        }
+
+        internal static SourceLinkMap Parse(string json, Action<string> reportDiagnostic)
+        {
+            bool errorReported = false;
+
+            void ReportInvalidJsonDataOnce(string message)
+            {
+                if (!errorReported)
+                {
+                    // Bad source link format
+                    reportDiagnostic($"The JSON format is invalid: {message}");
+                }
+
+                errorReported = true;
+            }
+
+            var list = new List<(FilePathPattern key, UriPattern value)>();
+            try
+            {
+                // trim BOM if present:
```

Comment:
nit:
```suggestion
                // trim UTF8 BOM if present:
```

Also: Doesn't Newtonsoft.Json handle a UTF8 BOM? @JamesNK?

Reply:
It uses StreamReader, so yes